crates/sui/src/framework_versions.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Context};
+use sui_framework_snapshot::{SingleSnapshot, FRAMEWORK_MANIFEST};
+use sui_protocol_config::ProtocolVersion;
+
+/// Return the framework snapshot for the latest known protocol version
+pub fn latest_framework() -> &'static SingleSnapshot {
+    FRAMEWORK_MANIFEST.last_key_value().expect("").1
+}
+
+/// Return the best commit hash for the given protocol version. Gives an error if [version]
+/// is newer than the maximum protocol version or older than the first known framework.
+pub fn framework_for_protocol(version: ProtocolVersion) -> anyhow::Result<&'static SingleSnapshot> {
+    if version > ProtocolVersion::MAX {
+        bail!("Protocol version {version:?} is newer than this CLI.");
```

Comment:
Maybe something more along the lines
```suggestion
        bail!("Protocol version {version:?} is newer than the one this binary was built with. Aborting.");
```

Reply:
Right. The code that calls this function will need to add the additional context to the error.

Reply:
Right, I did not check where this is being used to see how it's handled. Feel free to ignore!

However, `protocol version is newer than the one this CLI is aware / was built with` might make it more clear.

Reply:
> The reason not to do that is that there's a difference between gaps in the middle and gaps on the ends.

I think in both cases the correct behaviour is to return the greatest protocol version less than or equal to the queried one, right?

Reply:
There is no framework for version 0, so that should return an error for sure. For version 73 (which is existst but is not in the manifest), the version for 72 should be used with no warning to the user. For version 74 (which doesn't exist as far as this binary knows), 72 should still be used, but the user should be warned.

Reply:
Right, for 0 there should be an error, for everything else, the suggestion is that the function returns the protocol version it found and the associated sha, and then the caller gets to decide whether to emit a warning or not by comparing the protocol version it got with the one it asked for.

The benefit of doing that is that the code that emits the warning is closer to the other output code (it gets around the earlier tension about whether we can say the call will abort or not because of the error)

Reply:
so for `protocol_version(73)` (when 72 is the last entry in the manifest but the binary is built with 73), would you return `(72, sha)` or `(73, sha)`? What about for `protocol_version(57)` (55 and 58 are in the manifest but not 56 or 57)?

Reply:
FWIW with the current implementation, I expect the caller to look something like
```
let sha = framework_for_protocol(prot).or_else(|err|
    println!("[warning] couldn't find protocol version {prot}: {err}";
    latest_framework_version()
);
```

Reply:
Assuming the framework version table contains the following entries: `L..=55,_,_,58..=72`, I would expect a request for the framework at 73 to return the sha for 72, and a request for the framework at 57 to return the sha for 55. That matches what your suggested caller code snippet would do, but:

- Your suggested snippet would also end up picking the latest version if `framework_for_protocol` errored because the requested protocol was too low (like `0`) -- this seems unintended and not desirable.
- It gives you less control about when to produce the error message. Being one protocol version out of date is probably fine, and it may be annoying to badger people straight away in that case that their binary is out-of-date, when everything's probably fine (the only consequence of this issue is that there may be features in the new framework that they don't have access to, and the likelihood of that being the case for just one protocol version change seems ...low, I would probably wait like 4 or 8 versions or something -- that's 2 to 4 months worth of changes).